mm2src/kdf_walletconnect/src/session/rpc/settle.rs

```diff
@@ -0,0 +1,94 @@
+use crate::session::{Session, SessionProperties};
+use crate::storage::WalletConnectStorageOps;
+use crate::{error::WalletConnectError, WalletConnectCtxImpl};
+
+use common::log::{debug, info};
+use mm2_err_handle::prelude::{MapMmError, MmError, MmResult};
+use relay_rpc::domain::Topic;
+use relay_rpc::rpc::params::session_settle::SessionSettleRequest;
+
+pub(crate) async fn send_session_settle_request(
+    _ctx: &WalletConnectCtxImpl,
+    _session_info: &Session,
+) -> MmResult<(), WalletConnectError> {
+    // let mut settled_namespaces = BTreeMap::<String, Namespace>::new();
+    // let nam
+    // settled_namespaces.insert("eip155".to_string(), Namespace {
+    //     chains: Some(SUPPORTED_CHAINS.iter().map(|c| c.to_string()).collect()),
+    //     methods: SUPPORTED_METHODS.iter().map(|m| m.to_string()).collect(),
+    //     events: SUPPORTED_EVENTS.iter().map(|e| e.to_string()).collect(),
+    //     accounts: None,
+    // });
+    //
+    // let request = RequestParams::SessionSettle(SessionSettleRequest {
+    //     relay: session_info.relay.clone(),
+    //     controller: session_info.controller.clone(),
+    //     namespaces: SettleNamespaces(settled_namespaces),
+    //     expiry: Utc::now().timestamp() as u64 + THIRTY_DAYS,
+    //     session_properties: None,
+    // });
+    //
+    // ctx.publish_request(&session_info.topic, request).await?;
+
+    Ok(())
+}
+
+/// Process session settle request.
+pub(crate) async fn reply_session_settle_request(
+    ctx: &WalletConnectCtxImpl,
+    topic: &Topic,
+    settle: SessionSettleRequest,
+) -> MmResult<(), WalletConnectError> {
+    {
+        let mut session = ctx.session_manager.write();
+        let Some(session) = session.get_mut(topic) else {
+            return MmError::err(WalletConnectError::SessionError(format!("No session found for topic: {topic}")));
+        };
+        session.namespaces = settle.namespaces.0;
+        session.controller = settle.controller.clone();
+        session.relay = settle.relay;
+        session.expiry = settle.expiry;
+
+        if let Some(value) = settle.session_properties {
```

Comment:
`session_properties` field varies depending on the wallet and is currently used primarily by certain wallets like `Keplr`. Its fields can differ across wallets, as it isn’t standardized.

We can extend this using any enum once more wallets are discovered. Currently only `Keplr` sends this field during session initialization afaik.

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2223#discussion_r1897369080

Reply:
> any reason why not let `settle.session_properties` be of type `Option<SessionProperties>` instead of `Option<Value>`?

`settle.session_properties` is generic I mean it can be any arbitrary data depending on the wallet of course.

Reply:
so,
- if `session_properties` isn't sent, it's fine since it's optional,
- if `session_properties` is sent along in the request, we **MUST** succeed in parsing it (that means we must have an enumration of all the possible shapes it could be)
    - if we fail to de-serialize `session_properties` we will fail the request (the **MUST** part).

why do we need it as `Option<Value>` if we will fail the request if we fail to deserialize it?

Reply:
not all wallet will send this payload e.g metamask doesn't send but Kelpr does and maybe TrustWallet

Reply:
if it's not sent we will deserialize it as `None` (because `session_properties: Option<SessionProps>`) so it's fine and exactly the same as using `Option<Value>`. the only diff is we let serde deser the value instead of we desering it ourselves.

Reply:
another thing, WalletConnect lib shouldn't handle this data since it can't know what type it is so that's why I delegate the data deserialization to the lib user, which I assume is the right way.

LIBRARY: doesn't have any idea what data it is so session_properties: `Option<Value>`
USER: have an idea of type of data we'll get so `Option<SessionProps>`